File: Backend/src/main/java/com/sencarmarket/module/notification/mapper/NotificationMapper.java
package com.sencarmarket.module.notification.mapper;

import com.sencarmarket.module.notification.dto.NotificationResponse;
import com.sencarmarket.module.notification.dto.SignalementResponse;
import com.sencarmarket.module.notification.entity.Notification;
import com.sencarmarket.module.notification.entity.Signalement;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Mapper pour les conversions Notification/Signalement - Respecte SRP et DRY
 */
@Component
public class NotificationMapper {

    public NotificationResponse toNotificationResponse(Notification notification) {
        if (notification == null) {
            return null;
        }
        
        return NotificationResponse.builder()
                .id(notification.getId())
                .utilisateurId(notification.getUtilisateurId())
                .titre(notification.getTitre())
                .message(notification.getMessage())
                .type(notification.getType())
                .estLu(notification.getEstLu())
                .referenceId(notification.getReferenceId())
                .referenceType(notification.getReferenceType())
                .dateCreation(notification.getDateCreation())
                .dateLecture(notification.getDateLecture())
                .build();
    }

    public List<NotificationResponse> toNotificationResponseList(List<Notification> notifications) {
        return notifications.stream()
                .map(this::toNotificationResponse)
                .collect(Collectors.toList());
    }

    public SignalementResponse toSignalementResponse(Signalement signalement) {
        if (signalement == null) {
            return null;
        }
        
        return SignalementResponse.builder()
                .id(signalement.getId())
                .utilisateurId(signalement.getUtilisateurId())
                .typeEntite(signalement.getTypeEntite())
                .entiteId(signalement.getEntiteId())
                .motif(signalement.getMotif())
                .description(signalement.getDescription())
                .statutTraitement(signalement.getStatutTraitement())
                .actionAdmin(signalement.getActionAdmin())
                .adminId(signalement.getAdminId())
                .dateTraitement(signalement.getDateTraitement())
                .dateSignalement(signalement.getDateSignalement())
                .build();
    }

    public List<SignalementResponse> toSignalementResponseList(List<Signalement> signalements) {
        return signalements.stream()
                .map(this::toSignalementResponse)
                .collect(Collectors.toList());
    }
}

File: Backend/src/modules/notification/services/notification.mapper.ts
import { Injectable } from '@nestjs/common';

import { NotificationResponseDto } from '../dto/notification-response.dto';
import { SignalementResponseDto } from '../dto/signalement-response.dto';
import { NotificationRecord, SignalementRecord } from '../notification.models';

@Injectable()
export class NotificationMapper {
  toNotificationResponse(notification: NotificationRecord): NotificationResponseDto {
    return {
      id: notification.id,
      utilisateurId: notification.utilisateur_id,
      titre: notification.titre,
      message: notification.message,
      type: notification.type,
      estLu: notification.est_lu,
      referenceId: notification.reference_id,
      referenceType: notification.reference_type,
      dateCreation: notification.created_at,
      dateLecture: notification.date_lecture,
    };
  }

  toSignalementResponse(signalement: SignalementRecord): SignalementResponseDto {
    const nom = `${signalement.utilisateur?.prenom ?? ''} ${signalement.utilisateur?.nom ?? ''}`.trim();

    return {
      id: signalement.id,
      utilisateurId: signalement.utilisateur_id,
      utilisateurNom: nom || null,
      typeEntite: signalement.type_entite,
      entiteId: signalement.entite_id,
      motif: signalement.motif,
      description: signalement.description,
      statutTraitement: signalement.statut_traitement,
      actionAdmin: null,
      adminId: signalement.traite_par,
      dateTraitement: signalement.date_traitement,
      dateSignalement: signalement.created_at,
    };
  }
}
